22.2.8 release notes (#16758)
cockroachdb/docs: _includes/releases/v22.2/v22.2.8.md

## v22.2.8

Release Date: April 17, 2023

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v22-2-8-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

- [Backup schedules](../v22.2/manage-a-backup-schedule.html) created or altered to have the option `on_previous_running` will now have the full backup schedule created with the user specified option, but will override the incremental backup schedule to always default to `on_previous_running = wait`. This prevents duplicate incremental jobs from racing against each other, and ensures correctness of the backup chains created by the incremental schedule. [#98860][#98860]

> - [AVRO](../v22.2/changefeed-messages.html#avro) schema registry URI now allow additional `timeout=T` query parameter to change the default timeout for contacting schema registry. [#99505][#99505]

Reviewer: ```suggestion
- [AVRO](../v22.2/changefeed-messages.html#avro) schema registry URIs now allow you to change the default timeout for contacting the schema registry by setting the query parameter `timeout=T`. [#99505][#99505]
```

<h3 id="v22-2-8-sql-language-changes">SQL language changes</h3>

- Fixed the helper message on the [UPDATE](../v22.2/update.html) statement to correctly position the optional `FROM` clause. [#99299][#99299]
- Added a new `prepared_statements_cache_size` [cluster setting](../v22.2/cluster-settings.html) which, when set to a non-zero number of bytes, causes the least recently-used prepared statements to be automatically deallocated when prepared statement memory usage goes above the cache size. This setting can be used to avoid prepared statement leaks from long-lived connections which never `DEALLOCATE` prepared statements. [#99259][#99259]

> - Added two new [cluster settings](../v22.2/cluster-settings.html) that enable users to change the number of histogram samples and buckets collected when building histograms as part of table statistics collection: `sql.stats.histogram_samples.count` and `sql.stats.histogram_buckets.count`. While the default values should work for most cases, it may be beneficial to increase the number of samples and buckets for very large tables to avoid creating a histogram that misses important values. [#100544][#100544]

Reviewer: ```suggestion
- Added two new [cluster settings](../v22.2/cluster-settings.html) that enable users to change the number of histogram samples and buckets collected when building histograms as part of table statistics collection: `sql.stats.histogram_samples.count` and `sql.stats.histogram_buckets.count`. While the default values should work for most cases, it may be beneficial to increase the number of samples and buckets for very large tables to avoid creating a histogram that misses important values. [#100544][#100544]
```
I'm not sure the While...values bit belongs in a release note. WDYT?

Author: Removed.

<h3 id="v22-2-8-operational-changes">Operational changes</h3>

> - The `kv.trace.slow_request_stacks.threshold` [cluster setting](../v22.2/cluster-settings.html) can be used to attach available stack history from tracer snapshots to the trace of slow requests. [#99738][#99738]

Reviewer: I'm having some trouble parsing the last bit off this.

<h3 id="v22-2-8-db-console-changes">DB Console changes</h3>

- New data on the [Statement](../v22.2/ui-statements-page.html) and [Transaction](../v22.2/ui-transactions-page.html) pages is now automatically fetched every five minutes. [#99271][#99271]

> - On DB Console [Stats]() pages, you can now issue a new request for stats while a previous request is pending. The new request replaces the previous request. [#99271][#99271]

Reviewer 2: @lnhsingh `[Stats]()` needs a URL. This is causing htmltest to fail.

> - New data is now automatically fetched every 5 minutes on the [Statement](../v22.2/ui-statements-page.html) and [Transaction](../v22.2/ui-transactions-page.html) pages. [#99271][#99271]

Reviewer: ```suggestion
- New data on the [Statement](../v22.2/ui-statements-page.html) and [Transaction](../v22.2/ui-transactions-page.html) pages is now automatically fetched every five minutes. [#99271][#99271]
```

> - On DB console Stats pages, issuing a new request for stats while one is pending is now allowed and will replace the pending request. [#99271][#99271]

Reviewer: ```suggestion
- On DB Console [Stats]() pages, you can now issue a new request for stats while a previous request is pending. The new request replaces the previous request. [#99271][#99271]
```

<h3 id="v22-2-8-bug-fixes">Bug fixes</h3>

> - Prepared statements using placeholders in recursive [common table expressions](../v22.2/common-table-expressions.html) did not always re-optimize correctly after plugging in the parameters, which led to poor plan choices. This has been fixed. [#100327][#100327]

Reviewer: ```suggestion
- Fixed a bug where prepared statements using placeholders in recursive [common table expressions](../v22.2/common-table-expressions.html) did not always re-optimize after placeholders were resolved. [#100327][#100327]
```

- Fixed a bug where glob patterns that did not match tables in [`GRANT`](../v22.2/grant.html) or [`REVOKE`](../v22.2/revoke.html) statements would return an internal error with a confusing message instead of the `"no objects matched"` error. [#99437][#99437]

> - Since v22.2.0, CockroachDB could crash with the `"attempting to append refresh spans after the tracked timestamp has moved forward"` error in some rare cases (most likely when querying `pg_catalog` and `crdb_internal` virtual tables). This has been fixed. The workaround before upgrading would be to run `SET CLUSTER SETTING sql.distsql.use_streamer.enabled = false;`. [#99445][#99445]

Reviewer: ```suggestion
- Fixed a bug introduced in v22.2.0, where the node could crash with the error `"attempting to append refresh spans after the tracked timestamp has moved forward"` in some rare cases. To work around this issue until you can upgrade, run `SET CLUSTER SETTING sql.distsql.use_streamer.enabled = false;`. [#99445][#99445]
```

> - When using `cluster_logical_timestamp()` as a [`DEFAULT`](../v22.2/default-value.html) expression, the node would sometimes crash. This has been fixed.[#99662][#99662]

Reviewer: ```suggestion
- Fixed a bug where the node could crash when a statement used `cluster_logical_timestamp()` as a [`DEFAULT`](../v22.2/default-value.html) expression. [#99662][#99662]
```

> - The [`ALTER DEFAULT PRIVILEGES ... ON FUNCTIONS ...`](../v22.2/alter-default-privileges.html) statement is no longer allowed unless all nodes are running on v22.2 and the upgrade is finalized. This prevents a bug where executing this command in a mixed version cluster could cause nodes still running on v22.1 to crash. [#99845][#99845]

Reviewer: ```suggestion
- The [`ALTER DEFAULT PRIVILEGES ... ON FUNCTIONS ...`](../v22.2/alter-default-privileges.html) statement is no longer allowed unless all nodes are running on v22.2 and the upgrade is finalized.  This command could cause a node still running v22.1 to crash in a mixed-version cluster. [#99845][#99845]
```

> - Fixed a bug where the [`TRUNCATE TABLE`](../v22.2/truncate.html) GC job can be stuck in running status if the table descriptor has been GCed. This happened because `TRUNCATE TABLE` creates new empty indexes, then replaces and drops the old indexes. The dropped indexes data are deleted and GCed within the `TRUNCATE TABLE` GC job, which are needed to see the table descriptor make progress. However, if the table data has been GCed, the job couldn't make progress. This patch makes the GC job able to handle the missing descriptor edge case and let the `TRUNCATE TABLE` GC job succeed. [#100137][#100137]

Reviewer: This is hard to understand -- I tried an edit or two bug didn't make much progress. At the least it could use a copyedit, but it seems like maybe too much info for a RN.

- Fixed a bug that would cause the gateway node to crash if there are self-referencing [views](../v22.2/views.html). This bug is present since at least v21.1. [#100163][#100163]

> - Fixed a bug where queries reading from virtual tables (e.g. `crdb_internal.*` and `pg_catalog.*` tables) could hang indefinitely in the rare case when they resulted in an error. [#99968][#99968]

Reviewer: ```suggestion
- Fixed a bug where queries reading from virtual tables such as those in the `crdb_internal` and `pg_catalog` system catalogs could hang indefinitely if the query would result in an error. [#99968][#99968]
```

> - In rare cases involving overload and schema changes, users could sometimes, transiently, see errors of the form `"deadline below read timestamp is nonsensical; txn has would have no chance to commit"`. These errors carried an internal pgcode and could not be retried. This form of error is now classified as a retriable error and will be retried automatically either by the client or internally. [#100255][#100255]

Reviewer: "overload and schema changes" not sure what overload means here.

- Fixed a bug where the `sql.mem.distsql.current` metric would double count the memory usage of remote DistSQL flows. [#100254][#100254]
- Fixed a bug that could prevent a cached query with a [user-defined type](../v22.2/create-type.html) reference from being invalidated, even after a schema change that should prevent the type from being resolved. [#100223][#100223]
- Fixed a bug that could prevent a cached query from being invalidated when a [user-defined function](../v22.2/user-defined-functions.html) referenced by that query was altered or dropped. [#100223][#100223]
- Fixed a bug where [user-defined functions](../v22.2/user-defined-functions.html) were introduced that could cause a function call to resolve to the wrong function after changes to the schema search path. [#100223][#100223]
- Fixed a rare bug existing since before v22.1 that could cause a projected expression to replace column references with the wrong values. [#97593][#97593]

> - Fixed a bug where concurrent DML queries could corrupt an existing primary index when [adding a new column](../v22.2/alter-table.html) with a new column family. If a rollback occurs the table may no longer be accessible. [#100187][#100187]

Reviewer 3: The new text LGTM. Thanks for the edit @mdlinville!

<h3 id="v22-2-8-performance-improvements">Performance improvements</h3>

> - Removed prettify usages that could cause OOM on [Statements](../v22.2/ui-statements-page.html) and [Transactions](../v22.2/ui-transactions-page.html) page. [#99453][#99453]

Reviewer: ```suggestion
- Removed prettify usages that could cause out-of-memory (OOM) errors on the [Statements](../v22.2/ui-statements-page.html) and [Transactions](../v22.2/ui-transactions-page.html) pages. [#99453][#99453]
```

> - Audit logging could incur extra latency when resolving table/view/sequence names. To address this, this patch adds support for using the lease cache for these looks up and eliminate any extra round trips. [#99661][#99661]

Reviewer: ```suggestion
- Audit logging now looks up the names of tables, views, and sequences in a node's lease cache to reduce network latency. [#99661][#99661]
```

> - The execution of multiple [`FOREIGN KEY`](../v22.2/foreign-key.html) and [`UNIQUE`](../v22.2/unique.html) constraint checks can be parallelized in some cases. As a result, these checks can be completed faster, especially so in multi-region environments where the checks require cross-region reads. The feature is enabled by default on v23.1 release, and in order to enable it on v22.2, one must change the private `sql.distsql.parallelize_checks.enabled` [cluster setting](../v22.2/cluster-settings.html) to `true`. [#100520][#100520]

Reviewer: ```suggestion
- The execution of multiple [`FOREIGN KEY`](../v22.2/foreign-key.html) and [`UNIQUE`](../v22.2/unique.html) constraint checks can be parallelized in some cases to help them complete more quickly. faster, especially so in multi-region environments where the checks require cross-region reads. This feature is disabled by default. To enable it, set the private `sql.distsql.parallelize_checks.enabled` [cluster setting](../v22.2/cluster-settings.html) to `true`. [#100520][#100520]
```
My understanding is that we don't release note anything about multiregion. Can you verify this and also what is a "private" cluster setting?

Author: Do you mean "multi-tenant"? We do document MR, but not multi-tenant.

Verifying "private" cluster setting now.

- Fixed a bug where the `sql.mem.distsql.current` metric would count the memory usage of remote DistSQL flows twice. [#100254][#100254]
- Fixed a bug that could prevent a cached query with a [user-defined type](../v22.2/create-type.html) reference from being invalidated, even after a schema change that should prevent the type from being resolved. [#100223][#100223]
- Fixed a bug that could prevent a cached query from being invalidated when a [user-defined function](../v22.2/user-defined-functions.html) referenced by that query was altered or dropped. [#100223][#100223]
- Fixed a bug where [user-defined functions](../v22.2/user-defined-functions.html) were introduced that could cause a function call to resolve to the wrong function after changes to the schema search path. [#100223][#100223]
- Fixed a rare bug introduced before v22.1 that could cause a projected expression to replace column references with incorrect values. [#97593][#97593]

> - Fixed a bug in the new [declarative schema changer](https://www.cockroachlabs.com/docs/stable/online-schema-changes.html#declarative-schema-changer) that could cause primary index corruption when an `ALTER TABLE..ADD COLUMN` statement executed concurrently with an `UPDATE` or `INSERT` statement if the schema change failed and was rolled back. [#100187][#100187]

Reviewer: ```suggestion
- Fixed a bug in the new [declarative schema changer](../v22.2/online-schema-changes.html#declarative-schema-changer) that could cause primary index corruption when an `ALTER TABLE..ADD COLUMN` statement executed concurrently with an `UPDATE` or `INSERT` statement if the schema change failed and was rolled back. [#100187][#100187]
```

<h3 id="v22-2-8-performance-improvements">Performance improvements</h3>

- Removed prettify usages that could cause out-of-memory (OOM) errors on the [Statements](../v22.2/ui-statements-page.html) and [Transactions](../v22.2/ui-transactions-page.html) pages. [#99453][#99453]
- Audit logging now looks up the names of tables, views, and sequences in a node's lease cache to reduce network latency. [#99661][#99661]

> - The execution of multiple [`FOREIGN KEY`](../v22.2/foreign-key.html) and [`UNIQUE`](../v22.2/unique.html) constraint checks can be parallelized in some cases to help them complete more quickly. faster, especially so in multi-region environments where the checks require cross-region reads. This feature is disabled by default. To enable it, set the private (undocumented) `sql.distsql.parallelize_checks.enabled` [cluster setting](../v22.2/cluster-settings.html) to `true`. [#100520][#100520]

Reviewer: Are we OK with having this in a release note or should this be support-only since it is private?
